wallet/update.go

```diff
@@ -42,16 +59,58 @@ type (
 		SiafundStateElements() ([]types.StateElement, error)
 		UpdateSiafundStateElements([]types.StateElement) error
 
+		UpdateStateTree([]TreeNodeUpdate) error
+
 		AddressRelevant(types.Address) (bool, error)
 
 		ApplyIndex(types.ChainIndex, AppliedState) error
 		RevertIndex(types.ChainIndex, RevertedState) error
 	}
 )
 
+// updateStateElements updates the state elements in a store according to the
+// changes made by a chain update.
+func updateStateElements(tx UpdateTx, update stateTreeUpdater, indexMode IndexMode) error {
+	if indexMode == IndexModeFull {
+		var updates []TreeNodeUpdate
+		update.ForEachTreeNode(func(row, col uint64, h types.Hash256) {
+			updates = append(updates, TreeNodeUpdate{h, int(row), int(col)})
+		})
+		return tx.UpdateStateTree(updates)
+	} else {
+		// fetch all siacoin and siafund state elements
+		siacoinStateElements, err := tx.SiacoinStateElements()
+		if err != nil {
+			return fmt.Errorf("failed to get siacoin state elements: %w", err)
+		}
+
+		// update siacoin element proofs
+		for i := range siacoinStateElements {
+			update.UpdateElementProof(&siacoinStateElements[i])
+		}
+
+		if err := tx.UpdateSiacoinStateElements(siacoinStateElements); err != nil {
+			return fmt.Errorf("failed to update siacoin state elements: %w", err)
+		}
+
+		siafundStateElements, err := tx.SiafundStateElements()
+		if err != nil {
+			return fmt.Errorf("failed to get siafund state elements: %w", err)
+		}
+
+		// update siafund element proofs
+		for i := range siafundStateElements {
+			update.UpdateElementProof(&siafundStateElements[i])
+		}
+		return tx.UpdateSiafundStateElements(siafundStateElements)
+	}
```

Comment:
Can we add a sanity check here for `IndexModeNone`? I assume `updateStateElements` should never be called for `IndexModeNone`?